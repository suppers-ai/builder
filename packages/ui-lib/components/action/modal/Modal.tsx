import { BaseComponentProps } from "../../types.ts";

export interface ModalProps extends BaseComponentProps {
  open?: boolean;
  title?: string;
  backdrop?: boolean;
  responsive?: boolean;
  onClose?: () => void;
  onOpen?: () => void;
  onBackdropClick?: () => void;
}

export function Modal({
  children,
  class: className = "",
  open = false,
  title,
  backdrop = true,
  responsive = true,
  id,
  ...props
}: ModalProps) {
  if (!open) return null;

  const modalClasses = [
    "modal",
    open ? "modal-open" : "",
    className,
  ].filter(Boolean).join(" ");

  const modalBoxClasses = [
    "modal-box",
    responsive ? "w-11/12 max-w-5xl" : "",
  ].filter(Boolean).join(" ");

  return (
    <div class={modalClasses} id={id} {...props}>
      <div class={modalBoxClasses}>
        {title && <h3 class="font-bold text-lg mb-4">{title}</h3>}
        {children}
      </div>
      {backdrop && <div class="modal-backdrop"></div>}
    </div>
  );
}
