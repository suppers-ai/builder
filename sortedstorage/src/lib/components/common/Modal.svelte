<script lang="ts">
	import { X } from 'lucide-svelte';
	import { createEventDispatcher } from 'svelte';
	import { fade, scale } from 'svelte/transition';
	
	export let open = false;
	export let title = '';
	export let size: 'sm' | 'md' | 'lg' | 'xl' = 'md';
	export let closable = true;
	
	const dispatch = createEventDispatcher();
	
	$: sizeClass = {
		'sm': 'max-w-md',
		'md': 'max-w-lg',
		'lg': 'max-w-3xl',
		'xl': 'max-w-5xl'
	}[size];
	
	function handleClose() {
		if (closable) {
			open = false;
			dispatch('close');
		}
	}
	
	function handleKeydown(e: KeyboardEvent) {
		if (e.key === 'Escape' && closable) {
			handleClose();
		}
	}
</script>

<svelte:window on:keydown={handleKeydown} />

{#if open}
	<!-- Backdrop -->
	<div 
		class="fixed inset-0 bg-black bg-opacity-50 modal-backdrop"
		transition:fade={{ duration: 200 }}
		on:click={handleClose}
		role="presentation"
		aria-hidden="true"
	/>
	
	<!-- Modal -->
	<div 
		class="fixed inset-0 flex items-center justify-center p-4 modal-wrapper"
		transition:fade={{ duration: 200 }}
	>
		<div
			class="relative w-full {sizeClass} bg-white rounded-xl shadow-2xl overflow-hidden modal-dialog"
			transition:scale={{ duration: 200 }}
			on:click|stopPropagation
			role="dialog"
			aria-modal="true"
		>
				<!-- Header (only shown if title is provided) -->
				{#if title}
					<div class="modal-header flex items-center justify-between px-6 py-4 border-b border-gray-200 bg-gradient-to-r from-gray-50 to-white">
						<h2 class="text-xl font-semibold text-gray-900">
							{title}
						</h2>
						{#if closable}
							<button
								on:click={handleClose}
								class="p-1.5 rounded-lg bg-white border border-gray-200 hover:bg-gray-50 transition-colors shadow-sm"
							>
								<X class="w-5 h-5 text-gray-500" />
							</button>
						{/if}
					</div>
				{/if}
				
				<!-- Content -->
				<div class="modal-content">
					<slot />
				</div>
				
				<!-- Footer -->
				{#if $$slots.footer}
					<div class="px-6 py-4 border-t border-gray-200 bg-gray-50">
						<slot name="footer" />
					</div>
				{/if}
			</div>
		</div>
{/if}

<style>
	.modal-backdrop {
		z-index: 9998;
	}
	
	.modal-wrapper {
		z-index: 9999;
		pointer-events: none;
	}
	
	.modal-wrapper > * {
		pointer-events: auto;
	}
	
	.modal-dialog {
		z-index: 10000;
		position: relative;
	}
	
	.modal-header {
		z-index: 1;
		position: relative;
	}
	
	.modal-content {
		max-height: calc(80vh - 120px);
		overflow: visible;
		position: relative;
		z-index: 1;
	}
	
	.modal-content::-webkit-scrollbar {
		width: 6px;
	}
	
	.modal-content::-webkit-scrollbar-track {
		background: transparent;
	}
	
	.modal-content::-webkit-scrollbar-thumb {
		background: var(--color-tertiary, #f1e2dd);
		border-radius: 3px;
	}
	
	.modal-content::-webkit-scrollbar-thumb:hover {
		background: var(--color-accent, #a16d5b);
	}
</style>