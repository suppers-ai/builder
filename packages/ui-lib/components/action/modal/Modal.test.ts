import { assertEquals, assertStringIncludes } from "https://deno.land/std@0.224.0/assert/mod.ts";
import { assertSnapshot } from "https://deno.land/std@0.224.0/testing/snapshot.ts";
import { renderToString } from "preact-render-to-string";
import { h } from "preact";
import { Modal } from "./Modal.tsx";

Deno.test("Modal - basic rendering when closed", () => {
  const html = renderToString(Modal({
    children: "Test Modal Content",
    open: false,
  }));
  assertEquals(html, "");
});

Deno.test("Modal - basic rendering when open", () => {
  const html = renderToString(Modal({
    children: "Test Modal Content",
    open: true,
  }));
  assertStringIncludes(html, "modal");
  assertStringIncludes(html, "modal-open");
  assertStringIncludes(html, "modal-box");
  assertStringIncludes(html, "Test Modal Content");
});

Deno.test("Modal - with title", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    title: "Modal Title",
  }));
  assertStringIncludes(html, "Modal Title");
  assertStringIncludes(html, "font-bold text-lg mb-4");
});

Deno.test("Modal - without title", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
  }));
  assertEquals(html.includes("font-bold text-lg mb-4"), false);
});

Deno.test("Modal - with backdrop", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    backdrop: true,
  }));
  assertStringIncludes(html, "modal-backdrop");
});

Deno.test("Modal - without backdrop", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    backdrop: false,
  }));
  assertEquals(html.includes("modal-backdrop"), false);
});

Deno.test("Modal - responsive sizing", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    responsive: true,
  }));
  assertStringIncludes(html, "w-11/12 max-w-5xl");
});

Deno.test("Modal - non-responsive sizing", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    responsive: false,
  }));
  assertEquals(html.includes("w-11/12 max-w-5xl"), false);
});

Deno.test("Modal - custom class", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    class: "custom-modal-class",
  }));
  assertStringIncludes(html, "custom-modal-class");
});

Deno.test("Modal - with id", () => {
  const html = renderToString(Modal({
    children: "Content",
    open: true,
    id: "test-modal",
  }));
  assertStringIncludes(html, 'id="test-modal"');
});

Deno.test("Modal - complex children", () => {
  const html = renderToString(Modal({
    children: h("div", {}, [
      h("p", {}, "Modal paragraph"),
      h("button", { class: "btn" }, "Action"),
    ]),
    open: true,
    title: "Complex Modal",
  }));
  assertStringIncludes(html, "Modal paragraph");
  assertStringIncludes(html, "btn");
  assertStringIncludes(html, "Complex Modal");
});

Deno.test("Modal - all props combined", () => {
  const html = renderToString(Modal({
    children: "Full featured modal",
    open: true,
    title: "Full Modal",
    backdrop: true,
    responsive: true,
    class: "test-class",
    id: "full-modal",
  }));
  assertStringIncludes(html, "modal-open");
  assertStringIncludes(html, "Full Modal");
  assertStringIncludes(html, "modal-backdrop");
  assertStringIncludes(html, "w-11/12 max-w-5xl");
  assertStringIncludes(html, "test-class");
  assertStringIncludes(html, 'id="full-modal"');
});

// Snapshot tests
Deno.test("Modal - HTML snapshot when open", async (t) => {
  const html = renderToString(Modal({
    children: "Snapshot test content",
    open: true,
    title: "Snapshot Modal",
    backdrop: true,
  }));
  await assertSnapshot(t, html);
});

Deno.test("Modal - HTML snapshot when closed", async (t) => {
  const html = renderToString(Modal({
    children: "This should not render",
    open: false,
    title: "Hidden Modal",
  }));
  await assertSnapshot(t, html);
});

Deno.test("Modal - HTML snapshot complex", async (t) => {
  const html = renderToString(Modal({
    children: h("div", { class: "space-y-4" }, [
      h("p", { class: "text-base-content" }, "Modal description goes here."),
      h("div", { class: "modal-action" }, [
        h("button", { class: "btn" }, "Cancel"),
        h("button", { class: "btn btn-primary" }, "Save"),
      ]),
    ]),
    open: true,
    title: "Action Modal",
    responsive: true,
    backdrop: true,
    class: "custom-modal",
  }));
  await assertSnapshot(t, html);
});
