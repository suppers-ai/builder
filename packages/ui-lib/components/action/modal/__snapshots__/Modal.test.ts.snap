export const snapshot = {};

snapshot[`ModalDisplay - HTML snapshot when open 1`] = `'<div class="modal modal-open"><div class="modal-box w-11/12 max-w-5xl"><h3 class="font-bold text-lg mb-4">Snapshot Modal</h3>Snapshot test content</div><div class="modal-backdrop"></div></div>'`;

snapshot[`ModalDisplay - HTML snapshot when closed 1`] = `""`;

snapshot[`ModalDisplay - HTML snapshot complex 1`] = `'<div class="modal modal-open custom-modal"><div class="modal-box w-11/12 max-w-5xl"><h3 class="font-bold text-lg mb-4">Action Modal</h3><div class="space-y-4"><p class="text-base-content">Modal description goes here.</p><div class="modal-action"><button class="btn">Cancel</button><button class="btn btn-primary">Save</button></div></div></div><div class="modal-backdrop"></div></div>'`;
